x/zoneconcierge/module_ibc.go

```diff
@@ -0,0 +1,272 @@
+func (im IBCModule) OnRecvPacket(
+	ctx sdk.Context,
+	modulePacket channeltypes.Packet,
+	relayer sdk.AccAddress,
+) ibcexported.Acknowledgement {
+	var packetData types.ZoneconciergePacketData
+	if errProto := types.ModuleCdc.Unmarshal(modulePacket.GetData(), &packetData); errProto != nil {
+		im.keeper.Logger(ctx).Error("Failed to unmarshal packet data with protobuf", "error", errProto)
+		if errJSON := types.ModuleCdc.UnmarshalJSON(modulePacket.GetData(), &packetData); errJSON != nil {
+			im.keeper.Logger(ctx).Error("Failed to unmarshal packet data with JSON", "error", errJSON)
+			return channeltypes.NewErrorAcknowledgement(fmt.Errorf("cannot unmarshal packet data with protobuf (error: %v) or JSON (error: %v)", errProto, errJSON))
+		}
+	}
+
+	switch packet := packetData.Packet.(type) {
+	case *types.ZoneconciergePacketData_ConsumerSlashing:
+		err := im.keeper.HandleConsumerSlashing(ctx, modulePacket.DestinationPort, modulePacket.DestinationChannel, packet.ConsumerSlashing)
+		if err != nil {
+			return channeltypes.NewErrorAcknowledgement(err)
+		}
+		return channeltypes.NewResultAcknowledgement([]byte("Consumer slashing handled successfully"))
+	// Add other packet types here if needed
+	default:
+		errMsg := fmt.Sprintf("unrecognized %s packet type: %T", types.ModuleName, packet)
+		return channeltypes.NewErrorAcknowledgement(errorsmod.Wrap(sdkerrors.ErrUnknownRequest, errMsg))
+	}
+}
+
+// OnAcknowledgementPacket implements the IBCModule interface
+func (im IBCModule) OnAcknowledgementPacket(
+	ctx sdk.Context,
+	modulePacket channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	var ack channeltypes.Acknowledgement
+	// `x/wasm` uses both protobuf and json to encoded acknowledgement, so we need to try both here
+	// - for acknowledgment message with errors defined in `x/wasm`, it uses json
+	// - for all other acknowledgement messages, it uses protobuf
+	if errProto := types.ModuleCdc.Unmarshal(acknowledgement, &ack); errProto != nil {
+		im.keeper.Logger(ctx).Error("cannot unmarshal packet acknowledgement with protobuf", "error", errProto)
+		if errJson := types.ModuleCdc.UnmarshalJSON(acknowledgement, &ack); errJson != nil {
+			im.keeper.Logger(ctx).Error("cannot unmarshal packet acknowledgement with json", "error", errJson)
+			return errorsmod.Wrapf(sdkerrors.ErrUnknownRequest, "cannot unmarshal packet acknowledgement with protobuf (error: %v) or json (error: %v)", errProto, errJson)
+		}
+	}
+
+	// // TODO (Babylon): Dispatch and process packet
+	// switch packet := modulePacketData.Packet.(type) {
+	// default:
+	// 	errMsg := fmt.Sprintf("unrecognized %s packet type: %T", types.ModuleName, packet)
+	// 	return errorsmod.Wrap(sdkerrors.ErrUnknownRequest, errMsg)
+	// }
+
+	switch resp := ack.Response.(type) {
+	case *channeltypes.Acknowledgement_Result:
+		im.keeper.Logger(ctx).Info("received an Acknowledgement message", "result", string(resp.Result))
+		ctx.EventManager().EmitEvent(
```

Comment:
should we emit typed events?

Reply:
added TODO on emitting typed event for each `EmitEvent` occurence